File: backend/src/routes/optimize.rs
use axum::{Router, routing::post};

use crate::{
    handlers::{compress::compress_pdf, optimize::{add_page_numbers, add_watermark}},
    state::AppState,
};

pub fn routes() -> Router<AppState> {
    Router::<AppState>::new()
        .route("/rust-api/kompres", post(compress_pdf))
        .route("/rust-api/watermark", post(add_watermark))
        .route("/rust-api/page-numbers", post(add_page_numbers))
}

File: frontend-react/src/lib/api/optimize.ts
import {
  postFile,
  postFileWithText,
} from "./client";

export type CompressionQuality =
  | "high"
  | "medium"
  | "low";

/**
 * Kompres PDF.
 */
export function compressPdf(
  file: File,
  quality: CompressionQuality,
): Promise<Blob> {
  return postFileWithText(
    "/api/v1/pdf/compress",
    file,
    "quality",
    quality,
  );
}

/**
 * Menambahkan watermark ke PDF.
 */
export function addWatermark(
  file: File | Blob,
  text: string,
): Promise<Blob> {
  return postFileWithText(
    "/api/v1/pdf/watermark",
    file,
    "text",
    text,
  );
}

/**
 * Menambahkan nomor halaman ke PDF.
 */
export function addPageNumbers(
  file: File | Blob,
): Promise<Blob> {
  return postFile(
    "/api/v1/pdf/page-numbers",
    file,
  );
}
